ClientCore/Extensions/StringExtensions.cs

```diff
@@ -1,25 +1,77 @@
-        {
-            if (string.IsNullOrWhiteSpace(text))
-                return null;
-
-            int index = text.IndexOf("http://", StringComparison.Ordinal);
-            if (index == -1)
-                index = text.IndexOf("ftp://", StringComparison.Ordinal);
-            if (index == -1)
-                index = text.IndexOf("https://", StringComparison.Ordinal);
-
-            if (index == -1)
-                return null; // No link found
-
-            string link = text.Substring(index);
-            return link.Split(' ')[0]; // Nuke any words coming after the link
-        }
+        if (raw.Contains(ESCAPED_INI_NEWLINE_PATTERN, StringComparison.InvariantCulture))
+            throw new ArgumentException($"The string contains an illegal character sequence! ({ESCAPED_INI_NEWLINE_PATTERN})");
+
+        if (raw.Contains(ESCAPED_SEMICOLON, StringComparison.InvariantCulture))
+            throw new ArgumentException($"The string contains an illegal character sequence! ({ESCAPED_SEMICOLON})");
+
+        return raw
+            .Replace(ProgramConstants.INI_NEWLINE_PATTERN, ESCAPED_INI_NEWLINE_PATTERN, StringComparison.InvariantCulture)
+            .Replace(";", ESCAPED_SEMICOLON, StringComparison.InvariantCulture)
+            .Replace(Environment.NewLine, "\n", StringComparison.InvariantCulture)
+            .Replace("\n", ProgramConstants.INI_NEWLINE_PATTERN, StringComparison.InvariantCulture);
     }
-}
+
+    /// <summary>
+    /// Converts an INI-safe string to a normal string.
+    /// </summary>
+    /// <param name="iniString">Input INI string.</param>
+    /// <returns>Regular string.</returns>
+    public static string FromIniString(this string iniString)
+    {
+        return iniString
+            .Replace(ESCAPED_INI_NEWLINE_PATTERN, ProgramConstants.INI_NEWLINE_PATTERN, StringComparison.InvariantCulture)
+            .Replace(ESCAPED_SEMICOLON, ";", StringComparison.InvariantCulture)
+            .Replace(ProgramConstants.INI_NEWLINE_PATTERN, Environment.NewLine, StringComparison.InvariantCulture);
+    }
+
+    /// <summary>
+    /// Looks up a translated string for the specified key.
+    /// </summary>
+    /// <param name="defaultValue">The default string value as a fallback.</param>
+    /// <param name="key">The unique key name.</param>
+    /// <returns>The translated string value.</returns>
+    /// <remarks>
+    /// This method is referenced by <c>TranslationNotifierGenerator</c> in order to check if the const
+    /// values that are not initialized on client start automatically are missing (via notification
+    /// mechanism implemented down the call chain). Do not change the signature or move the method out
+    /// of the namespace it's currently defined in. If you do - you have to also edit the generator
+    /// source code to match.
+    /// </remarks>
+    public static string L10N(this string defaultValue, string key)
+        => Translation.Instance.LookUp(key, defaultValue);
```

Comment:
```suggestion
        => Translation.Instance?.LookUp(key, defaultValue) ?? defaultValue;
```

Prevents an additional exception when an exception occurs on startup before init has run.

Reply:
I am pretty sure we should prefer failsafe over failfast here. It shouldn't be called before initing a translation.

Reply:
It shouldn't but it can, PreStartup.HandleException() calls it. There are currently also L10N() calls in the code that runs before the init e.g.
`throw new NotImplementedException("-NOAUDIO is currently not implemented, please run the client without it.".L10N("Client:Main:NoAudio"));`

Reply:
In my opinion would be better to remove those instead.

Reply:
I prefer `Instance = new Translation()` instead of using `??`

Reply:
https://github.com/Metadorius/xna-cncnet-client/pull/9

How to re-open this conversation?